### PR TITLE
Fix on PHP tests on Windows

### PR DESCRIPTION
**Pull request by oyarzun**
- Only run PHP workflows when there are changes in the PHP cluster.
- Move Windows PHP tests to PHP workflow
- Fix and/or disable PHP test on Windows 

---
**^Add meaningful description above**

By opening a pull request you confirm that, unless explicitly stated otherwise, the changes -

 - are all your own work, and you have the right to contribute them.
 - are contributed solely under the terms and conditions of the Apache License 2.0 (see section 5 of the license for more information).

Please make sure (eg. `git log`) that all commits have a valid name and email address for you in the Author field.

If you're a first time contributor, see the Contributing guidelines for more information.

### CODE REVIEW

**Comment by junichi11** [comment]
I think the PHP area is affected by at least HTML, JavaScript, CSS, Editor, Project, and CSL (and more?).
Maybe, it's risky to run it only when there are changes in the PHP cluster.
Actually, in the past, something failed when there were changes in the JavaScript area (if I remember right)...

**Comment by JaroslavTulach** [comment]
Some PHP checks have to run on all commits. Could you split the tests into "fast" ~ 15min and "full"? Run the fast ones on every commit and "full" only:
- when there is a change in PHP files
- when a commit is integrated into `master` branch - e.g. post integration check

**Comment by neilcsmith-net** [comment]
A split sounds good, and potentially something to consider elsewhere too, but -

> when a commit is integrated into master branch - e.g. post integration check

- or possibly daily integration tests?
- maybe look at ensuring manually triggerable as well.
- most importantly, can't just be `master` - will need to cover `delivery` and `release***` branches too.

File: php/php.editor/test/unit/src/org/netbeans/modules/php/editor/verification/ImplementAbstractMethodsHintErrorTest.java [hunk 1]
```diff
@@ -169,6 +170,8 @@ public void testUnionTypes_04() throws Exception {
         checkHints(new ImplementAbstractMethodsHintError(), "testUnionTypesImplementMethod04.php");
     }
 
+    // fails on Windows
```
Comment (tmysik) [inline php/php.editor/test/unit/src/org/netbeans/modules/php/editor/verification/ImplementAbstractMethodsHintErrorTest.java:173]: I am not sure how this test could fail on Windows and pass on Linux... it is an editor test, so it is OS independent. Or am I missing something?

**Comment by tmysik** [comment]
The code changes in PHP are small, so no objections here (I just added one comment, not clear to me).

Unfortunately, I cannot comment on the other changes.

File: php/php.editor/test/unit/src/org/netbeans/modules/php/editor/verification/ImplementAbstractMethodsHintErrorTest.java (hunk 1 above)
Comment (oyarzun) [reply php/php.editor/test/unit/src/org/netbeans/modules/php/editor/verification/ImplementAbstractMethodsHintErrorTest.java:173]: I thought it was a due to the line endings, but changing the test file to use unix line ending still resulted in the following failure. https://github.com/oyarzun/netbeans/runs/5449321610?check_suite_focus=true#step:20:7291

```[junit] Testcase: testStaticReturnType_01(org.netbeans.modules.php.editor.verification.ImplementAbstractMethodsHintErrorTest):	FAILED
[junit] Content does not match between 'testfiles/verification/ImplementAbstractMethodsHintError/testStaticReturnTypeImplementMethod01.php' and 'testfiles/verification/ImplementAbstractMethodsHintError/testStaticReturnTypeImplementMethod01.phptestStaticReturnType_01.hints'
[junit] Expected content is:
[junit] 
[junit] class TestClass implements TestInterface {
[junit]       ---------
[junit] HINT:\TestClass is not abstract and does not override abstract method  testSimple() in \TestInterface
[junit] FIX:Implement All Abstract Methods
[junit] FIX:Declare Abstract Class
[junit] 
[junit] but actual is:
[junit] 
[junit] class TestClass implements TestInterface {
[junit]       ---------
[junit] HINT:\TestClass is not abstract and does not override abstract method  testNullableType() in \TestInterface
[junit] FIX:Implement All Abstract Methods
[junit] FIX:Declare Abstract Class
[junit] 
[junit] It differs in the following things:
[junit] 
[junit] Actual content contains following lines which are missing in expected content: 
[junit] 	HINT:\TestClass is not abstract and does not override abstract method  testNullableType() in \TestInterface
[junit] Expected content contains following lines which are missing in actual content: 
[junit] 	HINT:\TestClass is not abstract and does not override abstract method  testSimple() in \TestInterface
```

**Comment by oyarzun** [comment]
> I think the PHP area is affected by at least HTML, JavaScript, CSS, Editor, Project, and CSL (and more?).
> Maybe, it's risky to run it only when there are changes in the PHP cluster.
> Actually, in the past, something failed when there were changes in the JavaScript area (if I remember right)...

Initially, I had the following from taken from the php.editor dependencies. 

```yaml
paths:
  - .github/workflows/php.yml
  - php/**
  - platform/api.annotations.common/**
  - platform/api.progress/**
  - platform/api.progress.nb/**
  - platform/core.multiview/**
  - platform/editor.mimelookup/**
  - platform/openide.awt/**
  - platform/openide.dialogs/**
  - platform/openide.filesystems/**
  - platform/openide.loaders/**
  - platform/openide.modules/**
  - platform/openide.nodes/**
  - platform/openide.text/**
  - platform/openide.util/**
  - platform/openide.util.lookup/**
  - platform/openide.util.ui/**
  - platform/openide.windows/**
  - platform/options.api/**
  - platform/queries/**
  - ide/api.java.classpath/**
  - ide/csl.api/**
  - ide/csl.types/**
  - ide/db/**
  - ide/db.sql.editor/**
  - ide/editor/**
  - ide/editor.bracesmatching/**
  - ide/editor.codetemplates/**
  - ide/editor.document/**
  - ide/editor.fold/**
  - ide/editor.indent/**
  - ide/editor.lib/**
  - ide/editor.lib2/**
  - ide/editor.settings/**
  - ide/editor.util/**
  - ide/html.lexer/**
  - ide/lexer/**
  - ide/options.editor/**
  - ide/parsing.api/**
  - ide/parsing.indexing/**
  - ide/project.ant/**
  - ide/project.api/**
  - ide/projectuiapi.base/**
  - ide/spi.palette/**
  - ide/web.common/**
```

File: php/php.editor/test/unit/src/org/netbeans/modules/php/editor/verification/ImplementAbstractMethodsHintErrorTest.java (hunk 1 above)
Comment (tmysik) [reply php/php.editor/test/unit/src/org/netbeans/modules/php/editor/verification/ImplementAbstractMethodsHintErrorTest.java:173]: I see, thanks for clarification. Perhaps @junichi11 would know here?

**Comment by oyarzun** [comment]
> Some PHP checks have to run on all commits. Could you split the tests into "fast" ~ 15min and "full"? Run the fast ones on every commit and "full" only:
> 
> * when there is a change in PHP files
> * when a commit is integrated into `master` branch - e.g. post integration check
> 
> Of course it depends on what problem this PR is trying to solve? Unreliable PHP tests or slow PHP tests?

The initial intent was to avoid running the ~1.5 hours of PHP tests for every push, when the changes would not affect PHP. As  @junichi11 pointed out the PR as it stands does not cover potential dependencies. But in the process of changing the workflow, I found out that the PHP test on Windows were failing even though they the Windows workflow was showing passed.

The php.editor test suite take 1.5+ hours, the rest are fast.

File: php/php.editor/test/unit/src/org/netbeans/modules/php/editor/verification/ImplementAbstractMethodsHintErrorTest.java (hunk 1 above)
Comment (junichi11) [reply php/php.editor/test/unit/src/org/netbeans/modules/php/editor/verification/ImplementAbstractMethodsHintErrorTest.java:173]: Maybe the cause is `Set`. I'll try to improve it later (tomorrow or later).

**Comment by mbien** [comment]
> * Should we change the PR to use these paths (others needed)?
> 
> * and/or Add the ability to run the workflow manually?
> 
> * and/or Add a nightly schedule?

i think paths trigger for the long job + a short job (<30min) which runs every time would be a nice combination,

An alternative could be to always trigger everything like before, but make the long job conditional (e.g via a `integration-test` label):
https://docs.github.com/en/actions/using-workflows/triggering-a-workflow#example-using-a-value-in-the-event-payload

The PHP cluster is under active development, it think it is unlikely that bugs could sneak up on it since it isn't fully tested for a long time (modules in maintenance mode would be riskier). I am personally not a big fan of nightly schedules, its a last resort for special cases IMO.

**Comment by vieiro** [comment]
Hi,

It seems modules in the PHP cluster depend on (at least) these other 93 modules (possibly in other clusters).

I think we'd better build some tool to extract inter-module dependencies. This may be useful in the future if we want to explore different build systems or explore some degree of parallelization in builds, for instance.

As @junichi11  states, I think it's risky to disable PHP testing if we don't know **exactly** what the dependencies are. Note that there're many types of dependencies we need to take into account: build dependencies, test dependencies, runtime dependencies, etc.

I prefer to spend time in more unit-testing than spend time in resolving more issue reports.

Cheers,
Antonio

```
: com.jcraft.jsch
: org.netbeans.api.annotations.common
: org.netbeans.api.debugger
: org.netbeans.api.java.classpath
: org.netbeans.api.progress
: org.netbeans.api.progress.nb
: org.netbeans.api.search
: org.netbeans.api.templates
: org.netbeans.core.multiview
: org.netbeans.libs.commons_net
: org.netbeans.libs.javacup
: org.netbeans.libs.jsch.agentproxy
: org.netbeans.libs.json_simple
: org.netbeans.modules.code.analysis
: org.netbeans.modules.csl.api
: org.netbeans.modules.csl.types
: org.netbeans.modules.db
: org.netbeans.modules.db.kit
: org.netbeans.modules.db.sql.editor
: org.netbeans.modules.diff
: org.netbeans.modules.editor
: org.netbeans.modules.editor.bracesmatching
: org.netbeans.modules.editor.codetemplates
: org.netbeans.modules.editor.completion
: org.netbeans.modules.editor.document
: org.netbeans.modules.editor.fold
: org.netbeans.modules.editor.indent
: org.netbeans.modules.editor.indent.project
: org.netbeans.modules.editor.lib
: org.netbeans.modules.editor.lib2
: org.netbeans.modules.editor.mimelookup
: org.netbeans.modules.editor.settings
: org.netbeans.modules.editor.util
: org.netbeans.modules.extexecution
: org.netbeans.modules.extexecution.base
: org.netbeans.modules.gototest
: org.netbeans.modules.gsf.codecoverage
: org.netbeans.modules.gsf.testrunner
: org.netbeans.modules.gsf.testrunner.ui
: org.netbeans.modules.html.editor
: org.netbeans.modules.html.editor.lib
: org.netbeans.modules.html.lexer
: org.netbeans.modules.hudson
: org.netbeans.modules.hudson.ui
: org.netbeans.modules.keyring
: org.netbeans.modules.languages.apacheconf
: org.netbeans.modules.languages.ini
: org.netbeans.modules.lexer
: org.netbeans.modules.nbjunit
: org.netbeans.modules.options.api
: org.netbeans.modules.options.editor
: org.netbeans.modules.parsing.api
: org.netbeans.modules.parsing.indexing
: org.netbeans.modules.php.samples
: org.netbeans.modules.project.ant
: org.netbeans.modules.project.ant.ui
: org.netbeans.modules.projectapi
: org.netbeans.modules.projectuiapi
: org.netbeans.modules.projectuiapi.base
: org.netbeans.modules.queries
: org.netbeans.modules.refactoring.api
: org.netbeans.modules.selenium2
: org.netbeans.modules.selenium2.server
: org.netbeans.modules.spellchecker.apimodule
: org.netbeans.modules.web.browser.api
: org.netbeans.modules.web.clientproject.api
: org.netbeans.modules.web.common
: org.netbeans.modules.web.common.ui
: org.netbeans.modules.web.indent
: org.netbeans.modules.websvc.saas.api
: org.netbeans.modules.websvc.saas.codegen
: org.netbeans.modules.xml.jaxb.api
: org.netbeans.spi.debugger.ui
: org.netbeans.spi.editor.hints
: org.netbeans.spi.navigator
: org.netbeans.spi.palette
: org.netbeans.spi.viewmodel
: org.openide.actions
: org.openide.awt
: org.openide.dialogs
: org.openide.execution
: org.openide.explorer
: org.openide.filesystems
: org.openide.filesystems.nb
: org.openide.io
: org.openide.loaders
: org.openide.modules
: org.openide.nodes
: org.openide.text
: org.openide.util
: org.openide.util.lookup
: org.openide.util.ui
: org.openide.windows
```

File: php/php.editor/test/unit/src/org/netbeans/modules/php/editor/verification/ImplementAbstractMethodsHintErrorTest.java (hunk 1 above)
Comment (junichi11) [reply php/php.editor/test/unit/src/org/netbeans/modules/php/editor/verification/ImplementAbstractMethodsHintErrorTest.java:173]: @oyarzun Fixed: https://github.com/apache/netbeans/pull/3734

**Comment by oyarzun** [comment]
@junichi11 as it is too risky to only trigger the workflows when there are changes in the php cluster, I'm going to update this PR to just fix the Windows PHP tests after #3734 gets merged.

File: .github/workflows/windows.yml [hunk 2]
```diff
@@ -15,7 +15,7 @@
 # specific language governing permissions and limitations
 # under the License.
 
-name: Apache Netbeans
+name: Windows
```
Comment (mbien) [inline .github/workflows/windows.yml:18]: nitpick: i would keep the old name here until the other jobs are updated or update the "Check Linux" Job too for consistency.

**Comment by junichi11** [comment]
I don't have any objection if we can get notifications that tests failed for any changes PHP is affected. It would be great if we could reduce testing time or we could disable PHP tests properly. However, I don't have a good idea...

**Review by junichi11 — APPROVED** [review]

File: .github/workflows/windows.yml (hunk 2 above)
Comment (oyarzun) [reply .github/workflows/windows.yml:18]: @mbien I've updated the other workflow names and the badges in the readme.

Comment (mbien) [reply .github/workflows/windows.yml:18]: @oyarzun excellent

**Pull request merged**